### DISCUSSION TITLE
KeyError: 'footnote_ref': Subclass parser?

### DISCUSSION THREAD

**Discussion by nschloe** [executablebooks/meta#1104] [Q&A]
I'd like to use `MarkdownIt` to parse a Markdown document, make some changes (like a formatter),
and write it back to Markdown. I tried

```python
from markdown_it import MarkdownIt
from mdformat.renderer import MDRenderer
from mdit_py_plugins.front_matter import front_matter_plugin
from mdit_py_plugins.footnote import footnote_plugin


md = (
    MarkdownIt("gfm-like")
    .use(footnote_plugin)
)
text = """
[^a]

[^a]: Lorem
    ipsum.
"""

tokens = md.parse(text)

renderer = MDRenderer()
output_markdown = renderer.render(
    tokens,
    options={},
    env={},
)

print(output_markdown)
```

but this gives

```
KeyError: 'footnote_ref'
```

Apparently, the `MDRenderer` doesn't know footnotes, like the parser does. Would I have to subclass `MDRenderer` to add this functionality?

**Comment by chrisjsewell** [comment]
Heya, footnotes are an MD extension, for which there is an mdformat plugin. See the mdformat documentation for how to add/activate plugins

**Reply by nschloe** [reply]
Looking a https://mdformat.readthedocs.io/en/stable/users/plugins.html, I see that there are plugins. No idea how to use them with `MDRenderer()` though. Perhaps I'm looking at the wrong doc?
